### PR TITLE
Add support for slug fields in metaphysics v2

### PR DESCRIPTION
**Pull request by zephraph**
Will resolve PLATFORM-1552.

The purpose of this task is to remove `gravityID` from metaphysics v2 in favor of `slug` and `internalID` where appropriate. 

It requires manually combing through the list below to ensure the IDs are correct in MP v1 and match the below mapping for v2. 

- [x] "AggregationCount": { "id": "internalID" } 
- [x] "Artist": { "_id": "internalID", "id": "slug" }
- [x] "Artwork": { "_id": "internalID", "id": "slug" }
- [x] "ArtworkVersion": { "id": "internalID" }
- [x] "AuctionLot": { "id": "internalID" } (mapped to `AuctionResult`)
- [x] "Bidder": { "id": "internalID" }
- [x] "BidderPosition": { "id": "internalID" }
- [x] "BuyersPremium": { "_id": "internalID", "id": "slug" }
- [x] "Collection": { "id": "slug" }
- [x] "CollectorProfile": { "id": "internalID" }
- [x] "CreditCard": { "id": "internalID" }
- [x] "DaySchedule": { "_id": "internalID" }
- [x] "EditionSet": { "id": "internalID" }
- [x] "Fair": { "_id": "internalID", "id": "slug" }
- [x] "FairOrganizer": { "_id": "internalID", "id": "slug" }
- [x] "FeaturedLink": { "_id": "internalID", "id": "??? BOTH SAME" }
- [x] "Feedback": { "id": "internalID" }
- [x] "FollowArtist": { "id": "internalID" }
- [x] "FollowGene": { "id": "internalID" }
- [x] "FollowProfile": { "id": "internalID" }
- [x] "FollowShow": { "id": "internalID" }
- [x] "Gene": { "_id": "internalID", "id": "slug" }
- [x] "GeneFamily": { "_id": "internalID", "id": "slug" }
- [x] "GeneType": { "_id": "internalID", "id": "slug" }
- [x] "Image": { "id": "internalID" }
- [x] "InquiryRequest": { "id": "internalID" }
- [x] "Location": { "id": "internalID" }
- [x] "OrderedSet": { "_id": "internalID", "id": "??? BOTH SAME" }
- [x] "Partner": { "_id": "internalID", "id": "slug" }
- [x] "PartnerArtist": { "id": "internalID" }
- [x] "PartnerCategory": { "_id": "internalID", "id": "slug" }
- [x] "PartnerShow": { "_id": "internalID", "id": "slug" }t
- [x] "Profile": { "_id": "internalID", "id": "slug" }
- [x] "Sale": { "_id": "internalID", "id": "slug" }
- [x] "SaleArtwork": { "_id": "internalID", "id": "slug" }
- [x] "SiteHeroUnit": { "_id": "internalID", "id": "slug" }
- [x] "Tag": { "_id": "internalID", "id": "slug" }
- [x] "User": { "_id": "internalID", "id": "??? BOTH SAME" }

<details>
 <summary>Types not found in MP</summary>

- [ ] "AccessControl": { "id": "internalID" } 
- [ ] "AccountRequest": { "id": "internalID" }
- [ ] "AdditionalImage": { "id": "internalID" }
- [ ] "AdminNote": { "id": "internalID" }
- [ ] "AnonymousSession": { "id": "internalID" }
- [ ] "ArtistLocation": { "id": "slug" }
- [ ] "Bid": { "id": "internalID" }
- [ ] "ClearbitProfile": { "id": "internalID" }
- [ ] "CollectorReference": { "id": "internalID" }
- [ ] "CulturalMaker": { "_id": "internalID", "id": "slug" }
- [ ] "Culture": { "_id": "internalID" }
- [ ] "CustomerAccount": { "id": "internalID" }
- [ ] "Device": { "id": "internalID" }
- [ ] "FairAction": { "id": "internalID" }
- [ ] "FairEvent": { "id": "internalID" }
- [ ] "FairEventGroup": { "id": "internalID" }
- [ ] "FairMobileImage": { "id": "internalID" }
- [ ] "Feature": { "_id": "internalID", "id": "slug" }
- [ ] "Flag": { "id": "internalID" }
- [ ] "GeneAutomationRule": { "id": "internalID" }
- [ ] "InquiryRequestStatus": { "id": "internalID" }
- [ ] "LabFeature": { "id": "internalID" }
- [ ] "Map": { "id": "internalID" }
- [ ] "MerchantAccount": { "id": "internalID" }
- [x] "Offer": { "id": "internalID" }
- [ ] "OfferAction": { "id": "internalID" }
- [ ] "OfferCampaign": { "id": "internalID" }
- [ ] "OfferStatus": { "id": "internalID" }
- [ ] "Page": { "_id": "internalID", "id": "slug" }
- [ ] "PartnerAdminEvaluation": { "id": "internalID" }
- [ ] "PartnerAlbum": { "_id": "internalID", "id": "slug" }
- [ ] "PartnerArtistArtwork": { "id": "internalID" }
- [ ] "PartnerArtistDocument": { "id": "internalID" }
- [ ] "PartnerCommunication": { "_id": "internalID", "id": "slug" }
- [ ] "PartnerCommunicationContact": { "id": "internalID" }
- [ ] "PartnerContact": { "_id": "internalID", "id": "slug" }
- [ ] "PartnerLocation": { "id": "internalID" }
- [ ] "PartnerNetPromoterScore": { "id": "internalID" }
- [ ] "PartnerShowDocument": { "id": "slug" }
- [ ] "PartnerShowEvent": { "_id": "internalID" }
- [ ] "PartnerShowFairLocation": { "id": "internalID" }
- [ ] "PartnerShowImage": { "id": "internalID" }
- [ ] "PartnerSubscription": { "id": "internalID" }
- [ ] "PartnerSubscriptionCharge": { "id": "internalID" }
- [ ] "PartnerSubscriptionChargeLineItem": { "id": "internalID" }
- [ ] "PartnerSubscriptionChargeTransaction": { "id": "internalID" }
- [ ] "PartnerSubscriptionEvent": { "id": "internalID" }
- [ ] "PartnerSubscriptionPlan": { "id": "internalID" }
- [ ] "PartnerSubscriptionPlanComponent": { "id": "internalID" }
- [ ] "Period": { "_id": "internalID" }
- [ ] "ProfileCover": { "id": "internalID" }
- [ ] "ProfileIcon": { "id": "internalID" }
- [ ] "Purchase": { "id": "internalID" }
- [ ] "Shortcut": { "id": "internalID" }
- [ ] "SiteFeature": { "_id": "internalID", "id": "slug" }
- [ ] "SiteHeroUnitBackgroundImage": { "id": "internalID" }
- [ ] "SiteHeroUnitMobileBackgroundImage": { "id": "internalID" }
- [ ] "SiteHeroUnitRetinaTitleImage": { "id": "internalID" }
- [ ] "SiteHeroUnitTitleImage": { "id": "internalID" }
- [ ] "UserFairAction": { "id": "internalID" }
- [ ] "UserInterest": { "id": "internalID" }
- [ ] "UserSaleProfile": { "id": "internalID" }

</details>

## Types found in MP but no matching mapping

- HighestBid
- ArtworkLayers
- AttributionClass
- Article
- Author

## Types to clean up

- ArtworkContext* (replace w/ actual types i.e. sale)
- ArtworkFilter* (review for possible replacement w/ Tag)
- Highlighted*
- HomePageModuleContext*
- `Category` (Remove in V2 in favor of `PartnerCategory`)

(Go through all `{ "_id": "internalID" }` and manually verify if they are slugs or not)

### CODE REVIEW

**Comment by zephraph** [comment]
@alloy, I was looking at `ArtworkVersion` and before our v2 work it only had an `id` field (which maps over to a database id). We changed it to use `gravityID` which means it got `_id` and `__id` even though those weren't provided. Should all objects with some type of id also return an `__id` (so that it can be converted to `id` in the v2 transform)?

So `ArtworkVersion` would be changed to have `id` and `__id` only (which would ultimate be converted to `internalID` and `id` respectively).

**Comment by zephraph** [comment]
@alloy see 887ceef for the changes related to the above question. I went ahead and implemented it.

**Comment by zephraph** [comment]
Note that `Collection` already had a slug field.

**Comment by zephraph** [comment]
`DaySchedule` is special in that it doesn't have an `id`. This might require us to create a separate ID field definition similar to `GravityIDFields` but with a description that matches the `InternalIDField`. I'm gonna skip it for now.

**Comment by zephraph** [comment]
Note that updating `Fair` also updated `HomePageModuleContextFair`.

**Comment by zephraph** [comment]
`FairOrganizer` didn't have an id field. It's providing a `profile_id`. I'm just going to skip it for now because I'm not certain if I should actually introduce the id type.

File: _schema.graphql [hunk 1]
```diff
@@ -8475,6 +8502,14 @@ type OrderWithMutationSuccess {
...
 
 type organizer {
+  # A globally unique ID.
+  __id: ID!
+
+  # A slug ID.
+  id: ID!
+
+  # A type-specific ID likely used as a database ID.
+  _id: ID!
```
Comment (zephraph) [inline _schema.graphql]: These IDs weren't present originally, but they _are_ provided by [gravity](https://github.com/artsy/gravity/blob/5be396dc2722963af2ccccc924a9f5da563333f8/app/models/domain/fair_organizer.rb#L43)

**Comment by zephraph** [comment]
`PartnerShowEvent` is listed in MP v1 as `PartnerShowEventType`. I wonder if that was intended?

File: src/schema/v2/RenameFields.ts [hunk 2]
```diff
@@ -89,7 +89,9 @@ export class RenameFields implements Transform {
       if (newName) {
         if (this.changedFields[fieldKey(type, newName)] !== undefined) {
           throw new Error(
-            `Cannot rename two fields to the same name: ${newName}`
+            `Cannot rename two fields (${
+              this.changedFields[fieldKey(type, newName)]
+            }, ${oldName}) to the same name ${newName} on type ${type.name}`
```
Comment (alloy) [inline src/schema/v2/RenameFields.ts]: Nice, did you find these because of this error?

Comment (zephraph) [reply src/schema/v2/RenameFields.ts]: Yep! That error was helpful. I modified it so I could find the types easier though.

**Pull request merged**